Mathlib/Data/Ordering/Dickson.lean

```diff
@@ -0,0 +1,322 @@
+-- This means that the conclusion should take place
+-- in the quotient partial order associated with the preorder.
+theorem minimal_ne_and_finite_of {α : Type*} [PartialOrder α]
+    (H : ∀ a : ℕ → α, ∃ i j, i < j ∧ a i ≤ a j) (S : Set α) (hSne : S.Nonempty) :
+    let M := {x ∈ S | Minimal (fun x ↦ x ∈ S) x}
+    M.Nonempty ∧ M.Finite := by
+  constructor
+  · by_contra hM
+    rw [Set.not_nonempty_iff_eq_empty] at hM
+    by_cases hS : S.Finite
+    · -- in a finite set, there are minimal elements
+      obtain ⟨x, hx, hxm⟩ :=  Set.Finite.exists_minimal_wrt id S hS hSne
+      simp only [Set.sep_eq_empty_iff_mem_false] at hM
+      exact hM x hx (minimal_iff.mpr ⟨hx, hxm⟩)
+    · have : ∀ x : S, ∃ y : S, (y : α) < x := by
+        rintro ⟨x, hx⟩
+        simp only [Set.sep_eq_empty_iff_mem_false] at hM
+        by_contra hx'
+        push_neg at hx'
+        apply hM x hx
+        unfold Minimal
+        refine ⟨hx, ?_⟩
+        intro y hy
+        rw [le_iff_eq_or_lt]
+        rintro (hyx | hyx)
+        · exact le_of_eq hyx.symm
+        · exfalso
+          exact hx' ⟨y,hy⟩ hyx
+      obtain ⟨a0, ha0⟩ := hSne
+      let a : ℕ → S := Nat.rec ⟨a0, ha0⟩ fun _ x ↦ (this x).choose
+      have ha : ∀ n, (a (n + 1)).val < (a n).val := fun n ↦ (this (a n)).choose_spec
+      obtain ⟨i, j, H, H'⟩ := H (fun n ↦ (a n).val)
+      rw [← lt_self_iff_false (a i)]
+      exact lt_of_le_of_lt  H' (strictAnti_nat_of_succ_lt ha H)
+  · rw [← Set.not_infinite]
+    intro hM
+    obtain ⟨a, ha⟩ := Set.Infinite.natEmbedding _ hM
+    obtain ⟨i, j, h, H⟩ := H (fun n ↦ a n)
+    simp only [Set.mem_setOf_eq, Subtype.coe_le_coe] at H
+    exact ne_of_lt h <| ha <| le_antisymm H ((a j).prop.2.2 (a i).prop.1 H)
+
+-- Unless the equivalence classes for the preorder are finite,
+-- the assumption is often meaningless
+-- TODO : Generalize to `PartialOrder α`
+theorem isDickson_of_minimal_ne_and_finite
+    (H : ∀ (S : Set α) (_ : S.Nonempty), { x ∈ S | Minimal (fun x ↦ x ∈ S) x}.Nonempty
+        ∧ {x ∈ S | Minimal (fun x ↦ x ∈ S) x}.Finite) :
+    isDickson α := by
+  intro S
+  let B := {x ∈ S | Minimal (fun x ↦ x ∈ S) x}
+  have hBS : B ⊆ S := Set.sep_subset S (Minimal fun x ↦ x ∈ S)
+  use B
+  by_cases hS : S.Nonempty
+  · refine ⟨⟨hBS, ?_⟩, (H S hS).2⟩
+    intro a ha
+    let S' := {x ∈ S | x ≤ a}
+    have := H S' ⟨a, by simp [S', ha]⟩
+    obtain ⟨b, hb, hb'⟩ := this.1
+    refine ⟨b, ?_, hb.2⟩
+    simp only [B]
+    refine ⟨hb.1, ?_⟩
+    refine ⟨Set.mem_of_mem_inter_left hb, ?_⟩
+    intro y hy hyb
+    exact hb'.2 ⟨hy, le_trans hyb hb.2⟩ hyb
+  · rw [Set.not_nonempty_iff_eq_empty] at hS
+    have hB : B = ∅ := Set.subset_eq_empty hBS hS
+    constructor
+    exact ⟨hBS, by simp [hS]⟩
+    simp [hB, Finite.of_fintype]
+
+-- TODO : Generalize to `Preorder α`
+/-- Becker-Weispfenning, Proposition 4.42 -/
+theorem isDickson_tfae (α : Type*) [PartialOrder α] : List.TFAE [
+    isDickson α,
+    ∀ (a : ℕ → α), ∃ i j, i < j ∧ a i ≤ a j,
```

Comment:
This looks exactly like the definition of a partially well-ordered set. Indeed, this part of the tfae exactly matches a special case of `PartiallyWellOrderedOn`. What do you need from this definition that isn't already part of the existing one?

Reply:
For assumption 3, it's not that this set is finite, but that the equivalence classes (for the associated partial order) are finite. And when I proved that, I didn't think how to formalize the partial order associated to a preorder, or whether it already exists.

Reply:
In particular, this tfae gives you exactly the statement that `isDickson α` is equivalent to `PartiallyWellOrderedOn Set.univ (· ≤ ·)`, except that the latter is defined in greater generality. Additionally, using the API already given there gives you `isDickson_iff_exists_monotone` directly, instead of needing to write another proof.

Reply:
The general theorem proves the equivalence of three analogous properties for a preorder : being Dickson, property 2, and a variant of property 3 which I can't state in Lean but that asserts the corresponding finiteness property in the partial order associated with the preorder.

Reply:
Said otherwise : consider a preorder `X` and the partial order `Y` deduced from `X`. By definition, being Dickson is equivalent for `X` and `Y`, properties 1 and 2 are equivalent as well (because they involve only the strict relation), but property 3 is not. And it is for `Y` that it is equivalent to being Dickson.

Reply:
So if somebody really needs Dickson preorders, they could just prove the equivalence with being Dickson for the associated partial order, and apply the already proved theorem.

Reply:
> So if somebody really needs Dickson preorders, they could just prove the equivalence with being Dickson for the associated partial order, and apply the already proved theorem.

Dickson preorders, and Dickson partial orders are _already in mathlib_ however. This PR duplicates the definition, proves it is equivalent to the existing one, and re-proves lemmas which are already in mathlib. I'm afraid I still don't understand why it is useful to re-make the API, could you clarify?